remove tap arg

joemfb opened the pull request:
in order to preserve the crucial Strategic Parenthetical Reserves ...

for each `++in`, `++by`, and `++to`, i've replaced the `+-tap` implementation, indirecting through `+-rap` in case you want to keep the old name. The commits are granular, so it'll be easy to de/re-construct as needed ...

I've also done some dubious-quality work on the jets -- my latest version just dumps core. All this works without jet changes, though. I find that surprising ...

Finally, it's not based on the latest master, but a slightly older master: I couldn't get %clay to work on an up-to-date fake ~zod.

joemfb commented:
Also, I'm completely lost by the jet-compatibility hack in `+-tap:to`.

It's completely necessary: removing it produces

```
fine: parent not found (7016793f)
improper core 2a9080b6
improper label: [%tap 28265 %two %one 'k149' 0]
```

But how? There are no hints ...

juped commented:
Don't feel basing things on the latest `master` is necessary - that's a decadence which is reluctantly _permitted_, not a good practice :)

I think I know the deal with the jets, will let you know after I look

joemfb commented:
Here's what I've done so far: https://github.com/urbit/urbit/compare/master...joemfb:detap

again, that branch core dumps on startup (after a ~20 second delay)

juped commented:
That probably means you got a refcount wrong somewhere.

joemfb commented:
I just tweaked the existing `in_tap` jet to be argument-less, a la `in_wyt`, but obviously messed something up.

I did try refcounting the null here (since it's decremented later) https://github.com/joemfb/urbit/blob/5201eff5e7e66f4b76a1c4caa22bb15f0545f835/jets/d/in_tap.c#L31, but that didn't seem to help ...

something like

```
u3_noun b = u3_nul;
u3k(b);
```

juped commented:
Direct atoms aren't really refcounted and u3k/u3z are no-ops on them (though you should still get the refcounting semantics right on them just because), so that can't be it.

joemfb commented:
The only other thing I'm not sure about is the `u3r_mean`, but `u3x_con_2` (`+14`) seems intuitively to be the right axis.

Also, 

```
> =<  +2  |_  a/@  ++  b  |.(a)  --
[[1 0 14] 0 1]
```

cgyarvin commented:
Is this still stalled on the broken jet?  It looks to me like you need u3k(b) in your base case -- the arguments to _tap_in are retained, not transferred.

joemfb commented:
Sorry about this, I'm sure I'm just missing something obvious ...

I'm not sure what mean by the base case, though: the recursion at line 23? (https://github.com/joemfb/urbit/blob/5201eff5e7e66f4b76a1c4caa22bb15f0545f835/jets/d/in_tap.c#L23)

I've tried with and without u3k(b) there... no difference. FWIW, there isn't one in the current version: https://github.com/urbit/urbit/blob/master/jets/d/in_tap.c#L24

Here's a backtrace of the crash: http://pastebin.com/vZ4NL2C3. Does anything jump out?

cgyarvin commented:
I skimmed this too casually.  I'll look again later...

On Fri, Feb 24, 2017 at 5:09 PM, Joe Bryan <notifications@github.com> wrote:

> Sorry about this, I'm sure I'm just missing something obvious ...
>
> I'm not sure what mean by the base case, though: the recursion at line 23?
> (https://github.com/joemfb/urbit/blob/5201eff5e7e66f4b76a1c4caa22bb1
> 5f0545f835/jets/d/in_tap.c#L23)
>
> I've tried with and without u3k(b) there... no difference. FWIW, there
> isn't one in the current version: https://github.com/urbit/
> urbit/blob/master/jets/d/in_tap.c#L24
>
> Here's a backtrace of the crash: http://pastebin.com/vZ4NL2C3. Does
> anything jump out?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/urbit/arvo/pull/352#issuecomment-282446655>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AALyAS-KgGgltjjV0SbPXMZ5bTLyrTG3ks5rf39jgaJpZM4MBBKw>
> .
>

cgyarvin commented:
No, it'd just look weird.  It's also kind of abusing the type system.

The problem is also that I have a vague desire to someday extend auras to
all spans, not just atoms -- to do things like differentiating between a
set and a map.  Real runtime bugs happen because we don't have this
feature.  Also, it's retarded that jet matching is this broken.  (We also
need to put hash recognition back in.)

On Wed, Mar 15, 2017 at 1:49 PM, Joe Bryan <notifications@github.com> wrote:

> Ah, I see. I was interpreting the +-tap:to symptoms to be the reverse ...
> but that doesn't really make sense.
>
> This raises another question: all three of these tap/rap arms have the
> exact same tree-deconstruction implementation. Is there an advantage to
> repeating the implementation in the arms of each core? Would there be a
> potentially meaningful performance degradation if +-tap:by and +-tap:to
> where defined as ~(tap in a)?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/urbit/arvo/pull/352#issuecomment-286874441>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AALyAVTv6gbY3nTMX33-xPF6qIEG8a5-ks5rmE7NgaJpZM4MBBKw>
> .
>

joemfb commented:
This now works with the latest in urbit/urbit#820 (which has a link to a bootable pill).

Per my earlier question, is there any way I can jet these arms without what are effectively nested loops?

And I would assume `+-tap:to` should have no jet hint -- is that right? Any other changes I should make?

joemfb commented on sys/vane/ames.hoon:
```diff
@@ -432,7 +432,7 @@
       ?~  ack  [~ ~ liv]
       =.  ded  ?:(top [n.liv ded] ded)
-      =.  ded  ?:(vig.clu.u.ack (~(tap to r.liv) ded) ded)
+      =?  ded  vig.clu.u.ack  (weld ~(tap to r.liv) ded)
```
This is the one call site (along with a similar expression in `++flay` that had to be actually refactored (where I couldn't just remove parenthesis). The same expressions are repeated in`xmas.hoon`. `+-tap` used to prepend to its argument, so this isn't exactly a complex change. I just wanted to call it out specifically. I haven't tested the expression in a generator, but I haven't tested network operations ...

cgyarvin commented:
Good question -- you shouldn't need the extra |. there.  Instead, decompose
the |- into |. and =<  $.  Put the jet hint between the |. and the =< $.

I'm not super worried about +-tap:to.

Perhaps I will never understand why you changed the semantics of ++tap,
then added ++rap to implement the old semantics.  On the other hand, this
does force you to simplify all the code that taps with ~ -- a small heroic
effort.  Also if you miss a spot in a merge, the compiler won't be shy
about letting you know.

On Wed, Mar 15, 2017 at 7:52 PM, Joe Bryan <notifications@github.com> wrote:

> *@joemfb* commented on this pull request.
> ------------------------------
>
> In sys/vane/ames.hoon
> <https://github.com/urbit/arvo/pull/352#discussion_r106329545>:
>
> > @@ -432,7 +432,7 @@
>            [& $(liv l.liv)]
>        ?~  ack  [~ ~ liv]
>        =.  ded  ?:(top [n.liv ded] ded)
> -      =.  ded  ?:(vig.clu.u.ack (~(tap to r.liv) ded) ded)
> +      =?  ded  vig.clu.u.ack  (weld ~(tap to r.liv) ded)
>
> This is the one call site (along with a similar expression in ++flay that
> had to be actually refactored (where I couldn't just remove parenthesis).
> The same expressions are repeated inxmas.hoon. +-tap used to prepend to
> its argument, so this isn't exactly a complex change. I just wanted to call
> it out specifically. I haven't tested the expression in a generator, but I
> haven't tested network operations ...
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/urbit/arvo/pull/352#pullrequestreview-27245758>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AALyAZ7VquWNCmNcH9ewrgMEQG9-FV0nks5rmKPtgaJpZM4MBBKw>
> .
>

joemfb commented:
I implemented the new semantics in rap, changed all the call sites to use that, then renamed everything back to tap. My goal was always to remove the extra parens, but I wasn't sure how much of this would be approved -- there was some argument in the original issue. Since this is all going in (and a new pill/ivory.c is required), there's no good reason for the indirection. This PR (and urbit/urbit#820, for that matter) could be squashed into a single commit. Would you prefer that?

As you suggest, I originally had just one decomposed loop:

```
=<  $
~%  %tap  +  ~
=|  b/(list)
|.  ^+  b
```

but that axis is wrong and the jet wasn't being called. Correcting the axis to `+>` 
(or just using `~/  %tap`) didn't call the jet either ... The duplication I have in this PR is the only combination I could find that properly registered and called into the jet. Any idea why that might be?

cgyarvin commented:
You're trying to jet something that isn't a core.  The `~%` and the line
below it need to be inverted.  Sorry, this must have missed my attention
earlier.

As for commits, what's ideal is a sequence of steps to follow when merging,
each of which boots from the pill produced by double-compiling the last.
In rare cases we may produce a pill which cannot compile itself, only the
next stage.  This is to be avoided!

On Thu, Mar 16, 2017 at 6:45 PM, Joe Bryan <notifications@github.com> wrote:

> I implemented the new semantics in rap, changed all the call sites to use
> that, then renamed everything back to tap. My goal was always to remove the
> extra parens, but I wasn't sure how much of this would be approved -- there
> was some argument in the original issue. Since this is all going in (and a
> new pill/ivory.c is required), there's no good reason for the indirection.
> This PR (and urbit/urbit#820 <https://github.com/urbit/urbit/pull/820>,
> for that matter) could be squashed into a single commit. Would you prefer
> that?
>
> As you suggest, I originally had just one decomposed loop:
>
> =<  $
> ~%  %tap  +  ~
> =|  b/(list)
> |.  ^+  b
>
> but that axis is wrong and the jet wasn't being called. Correcting the
> axis to +>
> (or just using ~/ %tap) didn't call the jet either ... The duplication I
> have in this PR is the only combination I could find that properly
> registered and called into the jet. Any idea why that might be?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/urbit/arvo/pull/352#issuecomment-287244081>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AALyAaITZE_Ly-vozQRwTLVMmyUZV91Qks5rmeXUgaJpZM4MBBKw>
> .
>

ohAitch commented:
Is double-pilling really still necessary with the new boot sequence? Like,
if you edited hoon.hoon run +metal in test mode, but the second compile
will happen either way

On Thursday, 16 March 2017, cgyarvin <notifications@github.com
<javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:

> You're trying to jet something that isn't a core. The `~%` and the line
> below it need to be inverted. Sorry, this must have missed my attention
> earlier.
>
> As for commits, what's ideal is a sequence of steps to follow when merging,
> each of which boots from the pill produced by double-compiling the last.
> In rare cases we may produce a pill which cannot compile itself, only the
> next stage. This is to be avoided!
>
> On Thu, Mar 16, 2017 at 6:45 PM, Joe Bryan <notifications@github.com>
> wrote:
>
> > I implemented the new semantics in rap, changed all the call sites to use
> > that, then renamed everything back to tap. My goal was always to remove
> the
> > extra parens, but I wasn't sure how much of this would be approved --
> there
> > was some argument in the original issue. Since this is all going in (and
> a
> > new pill/ivory.c is required), there's no good reason for the
> indirection.
> > This PR (and urbit/urbit#820 <https://github.com/urbit/urbit/pull/820>,
> > for that matter) could be squashed into a single commit. Would you prefer
> > that?
> >
> > As you suggest, I originally had just one decomposed loop:
> >
> > =< $
> > ~% %tap + ~
> > =| b/(list)
> > |. ^+ b
> >
> > but that axis is wrong and the jet wasn't being called. Correcting the
> > axis to +>
> > (or just using ~/ %tap) didn't call the jet either ... The duplication I
> > have in this PR is the only combination I could find that properly
> > registered and called into the jet. Any idea why that might be?
> >
> > —
> > You are receiving this because your review was requested.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/urbit/arvo/pull/352#issuecomment-287244081>, or mute
> > the thread
> > <https://github.com/notifications/unsubscribe-auth/
> AALyAaITZE_Ly-vozQRwTLVMmyUZV91Qks5rmeXUgaJpZM4MBBKw>
> > .
> >
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/urbit/arvo/pull/352#issuecomment-287245988>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABxXhp9xgUYfWMHdn0xOOxZrw02TrV90ks5rmelTgaJpZM4MBBKw>
> .
>

cgyarvin commented:
I have to add a flag to make it not do the test compile.

Decrementing the Hoon version is also very helpful in these situations...

On Thu, Mar 16, 2017 at 7:03 PM, Anton Dyudin <notifications@github.com>
wrote:

> Is double-pilling really still necessary with the new boot sequence? Like,
> if you edited hoon.hoon run +metal in test mode, but the second compile
> will happen either way
>
> On Thursday, 16 March 2017, cgyarvin <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:
>
> > You're trying to jet something that isn't a core. The `~%` and the line
> > below it need to be inverted. Sorry, this must have missed my attention
> > earlier.
> >
> > As for commits, what's ideal is a sequence of steps to follow when
> merging,
> > each of which boots from the pill produced by double-compiling the last.
> > In rare cases we may produce a pill which cannot compile itself, only the
> > next stage. This is to be avoided!
> >
> > On Thu, Mar 16, 2017 at 6:45 PM, Joe Bryan <notifications@github.com>
> > wrote:
> >
> > > I implemented the new semantics in rap, changed all the call sites to
> use
> > > that, then renamed everything back to tap. My goal was always to remove
> > the
> > > extra parens, but I wasn't sure how much of this would be approved --
> > there
> > > was some argument in the original issue. Since this is all going in
> (and
> > a
> > > new pill/ivory.c is required), there's no good reason for the
> > indirection.
> > > This PR (and urbit/urbit#820 <https://github.com/urbit/urbit/pull/820
> >,
> > > for that matter) could be squashed into a single commit. Would you
> prefer
> > > that?
> > >
> > > As you suggest, I originally had just one decomposed loop:
> > >
> > > =< $
> > > ~% %tap + ~
> > > =| b/(list)
> > > |. ^+ b
> > >
> > > but that axis is wrong and the jet wasn't being called. Correcting the
> > > axis to +>
> > > (or just using ~/ %tap) didn't call the jet either ... The duplication
> I
> > > have in this PR is the only combination I could find that properly
> > > registered and called into the jet. Any idea why that might be?
> > >
> > > —
> > > You are receiving this because your review was requested.
> > > Reply to this email directly, view it on GitHub
> > > <https://github.com/urbit/arvo/pull/352#issuecomment-287244081>, or
> mute
> > > the thread
> > > <https://github.com/notifications/unsubscribe-auth/
> > AALyAaITZE_Ly-vozQRwTLVMmyUZV91Qks5rmeXUgaJpZM4MBBKw>
> > > .
> > >
> >
> > —
> > You are receiving this because you are subscribed to this thread.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/urbit/arvo/pull/352#issuecomment-287245988>, or mute
> > the thread
> > <https://github.com/notifications/unsubscribe-auth/
> ABxXhp9xgUYfWMHdn0xOOxZrw02TrV90ks5rmelTgaJpZM4MBBKw>
> > .
> >
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/urbit/arvo/pull/352#issuecomment-287246335>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AALyAduSDntz1YWkvmQp0csviiSf3hcOks5rmenkgaJpZM4MBBKw>
> .
>

joemfb commented:
I've tried both of these, neither calls into the jet:

```
=|  b/(list _?>(?=(^ a) n.a))
=<  $
~%  %rap  +>  ~
|.  ^+  b
```

```
=<  $
=|  b/(list _?>(?=(^ a) n.a))
~%  %rap  +>  ~
|.  ^+  b
```

The hint axis is relative to the core, right? It's the axis of the containing core, from the perspective of the jetted arm?

ohAitch commented:
I think you might want ~% %rap + ~ since it doesn't have a sample

On Thursday, 16 March 2017, Joe Bryan <notifications@github.com> wrote:

> I've tried both of these, neither calls into the jet:
>
> =|  b/(list _?>(?=(^ a) n.a))
> =<  $
> ~%  %rap  +>  ~
> |.  ^+  b
>
> =<  $
> =|  b/(list _?>(?=(^ a) n.a))
> ~%  %rap  +>  ~
> |.  ^+  b
>
> The hint axis is relative to the core, right? It's the axis of the
> containing core, from the perspective of the jetted arm?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/urbit/arvo/pull/352#issuecomment-287251587>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABxXhnTDpx1k3QCFW-RsBUkMXpNtLFk4ks5rmfMkgaJpZM4MBBKw>
> .
>

cgyarvin commented:
Anton, ~% annotates the product, not the subject.

More precisely, the axis (here ->) is the axis at which the next core up in
the stack is found, within the current core.  Arms are not at issue.  In
fact, `~/` should work here.

So I can't explain why this wouldn't work.  Of course that's normal when
debugging.  Perhaps a jet registration problem.

On Thu, Mar 16, 2017 at 7:43 PM, Anton Dyudin <notifications@github.com>
wrote:

> I think you might want ~% %rap + ~ since it doesn't have a sample
>
> On Thursday, 16 March 2017, Joe Bryan <notifications@github.com> wrote:
>
> > I've tried both of these, neither calls into the jet:
> >
> > =| b/(list _?>(?=(^ a) n.a))
> > =< $
> > ~% %rap +> ~
> > |. ^+ b
> >
> > =< $
> > =| b/(list _?>(?=(^ a) n.a))
> > ~% %rap +> ~
> > |. ^+ b
> >
> > The hint axis is relative to the core, right? It's the axis of the
> > containing core, from the perspective of the jetted arm?
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/urbit/arvo/pull/352#issuecomment-287251587>, or mute
> > the thread
> > <https://github.com/notifications/unsubscribe-auth/ABxXhnTDpx1k3QCFW-
> RsBUkMXpNtLFk4ks5rmfMkgaJpZM4MBBKw>
> > .
> >
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/urbit/arvo/pull/352#issuecomment-287251705>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AALyAYgV_sGcVuCgH8wE8Sa-5AqbT21Nks5rmfNVgaJpZM4MBBKw>
> .
>

joemfb commented:
I think I've confirmed that this is a dispatch problem, not a registration problem. I'll see if I can debug the dispatch.

This is most I've ever done will pills and breaching, and I'm not sure what you mean by double-compiling. I did apply these commits incrementally while developing and testing the jets, but they can be applied with a single pill.

- new fake-zod from master
- `|mount %`, `|autoload`
- checkout this arvo branch
- copy in `sys/*`
- `.tap/pill +metal ~zod`
- checkout and make urbit/urbit#820
- new fake-zod from tap.pill

So I don't see the utility of separate commits, now that everything works. Any objections to squashing this down to one?

cgyarvin commented:
I have no objections to squashing this down to one.  Just bear in mind that
we're doing this on master because there's no way to

I remain curious as to what's going in with the jet dispatch...

On Fri, Mar 17, 2017 at 7:38 AM, Joe Bryan <notifications@github.com> wrote:

> I think I've confirmed that this is a dispatch problem, not a registration
> problem. I'll see if I can debug the dispatch.
>
> This is most I've ever done will pills and breaching, and I'm not sure
> what you mean by double-compiling. I did apply these commits incrementally
> while developing and testing the jets, but they can be applied with a
> single pill.
>
>    - new fake-zod from master
>    - |mount %, |autoload
>    - checkout this arvo branch
>    - copy in sys/*
>    - .tap/pill +metal ~zod
>    - checkout and make urbit/urbit#820
>    <https://github.com/urbit/urbit/pull/820>
>    - new fake-zod from tap.pill
>
> So I don't see the utility of separate commits, now that everything works.
> Any objections to squashing this down to one?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/urbit/arvo/pull/352#issuecomment-287371560>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AALyAS6E1QSpl52Sm3Hnh_vASPz4PxbCks5rmprjgaJpZM4MBBKw>
> .
>

joemfb commented:
Looks like that got truncated -- there's no way to what?

cgyarvin commented:
Doh!  No way to do that over the air on maintenance...

Sent from my iPhone

> On Mar 17, 2017, at 11:24 AM, Joe Bryan <notifications@github.com> wrote:
> 
> Looks like that got truncated -- there's no way to what?
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

juped commented:
Please leave it unsquashed (or at least be judicious and just combine the separate in/by/to commits into one each)

joemfb commented:
@juped, I'm happy to structure this into any arbitrary set of commits that makes things easier for you -- just say the word. It seems like the indirection through `rap` doesn't serve any purpose, so these could be split by containing core (as you suggest), and then maybe again by kernel vs userspace?

joemfb commented:
I spent a bunch of time with `jets.c` and `runtime.md` this weekend, working through exactly how jet registration and dispatch work. I was making a bunch of assumptions which turned out not to be case.

The issue I was having was simply that my `rap` arm and jet where matching the already existing `tap` arm and jet. As we've already discussed `|=(^ .)` is exactly equivalent to `=|(^ |.(.))`, and jets are dispatched by the hash of the battery ...

Looking back over this discussion, there was a big clue in my original comment: "all of this works without jet changes ...". (I don't remember exactly what I was thinking; I guess I assumed that the arm was being naively evaluated as nock.) So this has been another episode in "joe does something stupid, kicks off a big discussion, and bangs his head against the wall until something falls out". Story of my career ...

I've updated urbit/urbit#820 so that it no longer touches `ivory.c`, and simply refactors the tap jet calling conventions. But it's completely independent of this PR -- this could even go out over the air, if need be.

juped commented:
I thought you were having these problems at the end, after existing tap was removed?

joemfb commented:
I've done a pretty terrible job communicating throughout this PR. 

I did my initial testing with both `+-tap` and `+-rap` present. There's an explanation for this, but, as I started to write it up, I realized just how bad it was. This is all somewhat embarrassing...

I was debugging via printfs in the jet, and getting utterly inundated with them, since everything calls tap. Of course, I could've just put the printf in `wdi_tap`, which is never directly called from C.

juped commented:
This seems like a rite of passage for a lot of people. Probably just need an extra paragraph in u3.md emphasizing how it works.

joemfb commented:
Well, I'm sorry for kicking off a wild goose chase. I would've preferred to pass thru the rite in private ... I'll try to digest what I've learned here and summarize it in a forum post.

BTW, the hint doesn't have to immediately precede the jetted core. The tail of the hint formula is evaluated before being passed to `u3j_mine`, so this works fine:

```
~/  %tap
=|  (list)
|.  ...
```
